File: src/WellTool.AI/hutool-ai/src/main/java/cn/hutool/ai/model/ollama/OllamaServiceImpl.java
/*
 * Copyright (c) 2025 Hutool Team and hutool.cn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.hutool.ai.model.ollama;

import cn.hutool.ai.AIException;
import cn.hutool.ai.core.AIConfig;
import cn.hutool.ai.core.BaseAIService;
import cn.hutool.ai.core.Message;
import cn.hutool.core.bean.BeanPath;
import cn.hutool.core.thread.ThreadUtil;
import cn.hutool.core.util.StrUtil;
import cn.hutool.http.Header;
import cn.hutool.http.HttpRequest;
import cn.hutool.http.HttpResponse;
import cn.hutool.json.JSONObject;
import cn.hutool.json.JSONUtil;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

/**
 * Ollama服务，AI具体功能的实现
 *
 * @author yangruoyu-yumeisoft
 * @since 5.8.40
 */
public class OllamaServiceImpl extends BaseAIService implements OllamaService {

	// 对话补全
	private static final String CHAT_ENDPOINT = "/api/chat";
	// 文本生成
	private static final String GENERATE_ENDPOINT = "/api/generate";
	// 文本嵌入
	private static final String EMBEDDINGS_ENDPOINT = "/api/embeddings";
	// 列出模型
	private static final String LIST_MODELS_ENDPOINT = "/api/tags";
	// 显示模型信息
	private static final String SHOW_MODEL_ENDPOINT = "/api/show";
	// 拉取模型
	private static final String PULL_MODEL_ENDPOINT = "/api/pull";
	// 删除模型
	private static final String DELETE_MODEL_ENDPOINT = "/api/delete";
	// 复制模型
	private static final String COPY_MODEL_ENDPOINT = "/api/copy";

	/**
	 * 构造函数
	 *
	 * @param config AI配置
	 */
	public OllamaServiceImpl(final AIConfig config) {
		super(config);
	}

	@Override
	public String chat(final List<Message> messages) {
		final String paramJson = buildChatRequestBody(messages);
		final HttpResponse response = sendPost(CHAT_ENDPOINT, paramJson);
		JSONObject responseJson = JSONUtil.parseObj(response.body());
		Object errorMessage = BeanPath.create("error").get(responseJson);
		if(errorMessage!=null){
			throw new RuntimeException(errorMessage.toString());
		}
		return BeanPath.create("message.content").get(responseJson).toString();
	}

	@Override
	public void chat(final List<Message> messages, final Consumer<String> callback) {
		Map<String, Object> paramMap = buildChatStreamRequestBody(messages);
		ThreadUtil.newThread(() -> sendPostStream(CHAT_ENDPOINT, paramMap, callback::accept), "ollama-chat-sse").start();
	}

	@Override
	public String generate(String prompt) {
		final String paramJson = buildGenerateRequestBody(prompt, null);
		final HttpResponse response = sendPost(GENERATE_ENDPOINT, paramJson);
		return response.body();
	}

	@Override
	public void generate(String prompt, Consumer<String> callback) {
		Map<String, Object> paramMap = buildGenerateStreamRequestBody(prompt, null);
		ThreadUtil.newThread(() -> sendPostStream(GENERATE_ENDPOINT, paramMap, callback::accept), "ollama-generate-sse").start();
	}

	@Override
	public String generate(String prompt, String format) {
		final String paramJson = buildGenerateRequestBody(prompt, format);
		final HttpResponse response = sendPost(GENERATE_ENDPOINT, paramJson);
		return response.body();
	}

	@Override
	public void generate(String prompt, String format, Consumer<String> callback) {
		Map<String, Object> paramMap = buildGenerateStreamRequestBody(prompt, format);
		ThreadUtil.newThread(() -> sendPostStream(GENERATE_ENDPOINT, paramMap, callback::accept), "ollama-generate-sse").start();
	}

	@Override
	public String embeddings(String prompt) {
		final String paramJson = buildEmbeddingsRequestBody(prompt);
		final HttpResponse response = sendPost(EMBEDDINGS_ENDPOINT, paramJson);
		return response.body();
	}

	@Override
	public String listModels() {
		final HttpResponse response = sendGet(LIST_MODELS_ENDPOINT);
		return response.body();
	}

	@Override
	public String showModel(String modelName) {
		final String paramJson = buildShowModelRequestBody(modelName);
		final HttpResponse response = sendPost(SHOW_MODEL_ENDPOINT, paramJson);
		return response.body();
	}

	@Override
	public String pullModel(String modelName) {
		final String paramJson = buildPullModelRequestBody(modelName);
		final HttpResponse response = sendPost(PULL_MODEL_ENDPOINT, paramJson);
		return response.body();
	}

	@Override
	public String deleteModel(String modelName) {
		final String paramJson = buildDeleteModelRequestBody(modelName);
		final HttpResponse response = sendDeleteRequest(DELETE_MODEL_ENDPOINT, paramJson);
		return response.body();
	}

	@Override
	public String copyModel(String source, String destination) {
		final String paramJson = buildCopyModelRequestBody(source, destination);
		final HttpResponse response = sendPost(COPY_MODEL_ENDPOINT, paramJson);
		return response.body();
	}

	// 构建chat请求体
	private String buildChatRequestBody(final List<Message> messages) {
		final Map<String, Object> paramMap = new HashMap<>();
		paramMap.put("stream",false);
		paramMap.put("model", config.getModel());
		paramMap.put("messages", messages);
		// 合并其他参数
		paramMap.putAll(config.getAdditionalConfigMap());

		return JSONUtil.toJsonStr(paramMap);
	}

	// 构建chatStream请求体
	private Map<String, Object> buildChatStreamRequestBody(final List<Message> messages) {
		final Map<String, Object> paramMap = new HashMap<>();
		paramMap.put("stream", true);
		paramMap.put("model", config.getModel());
		paramMap.put("messages", messages);
		// 合并其他参数
		paramMap.putAll(config.getAdditionalConfigMap());

		return paramMap;
	}

	// 构建generate请求体
	private String buildGenerateRequestBody(final String prompt, final String format) {
		final Map<String, Object> paramMap = new HashMap<>();
		paramMap.put("model", config.getModel());
		paramMap.put("prompt", prompt);
		if (StrUtil.isNotBlank(format)) {
			paramMap.put("format", format);
		}
		// 合并其他参数
		paramMap.putAll(config.getAdditionalConfigMap());

		return JSONUtil.toJsonStr(paramMap);
	}

	// 构建generateStream请求体
	private Map<String, Object> buildGenerateStreamRequestBody(final String prompt, final String format) {
		final Map<String, Object> paramMap = new HashMap<>();
		paramMap.put("stream", true);
		paramMap.put("model", config.getModel());
		paramMap.put("prompt", prompt);
		if (StrUtil.isNotBlank(format)) {
			paramMap.put("format", format);
		}
		// 合并其他参数
		paramMap.putAll(config.getAdditionalConfigMap());

		return paramMap;
	}

	// 构建embeddings请求体
	private String buildEmbeddingsRequestBody(final String prompt) {
		final Map<String, Object> paramMap = new HashMap<>();
		paramMap.put("model", config.getModel());
		paramMap.put("prompt", prompt);
		// 合并其他参数
		paramMap.putAll(config.getAdditionalConfigMap());

		return JSONUtil.toJsonStr(paramMap);
	}

	// 构建showModel请求体
	private String buildShowModelRequestBody(final String modelName) {
		final Map<String, Object> paramMap = new HashMap<>();
		paramMap.put("name", modelName);

		return JSONUtil.toJsonStr(paramMap);
	}

	// 构建pullModel请求体
	private String buildPullModelRequestBody(final String modelName) {
		final Map<String, Object> paramMap = new HashMap<>();
		paramMap.put("name", modelName);

		return JSONUtil.toJsonStr(paramMap);
	}

	// 构建deleteModel请求体
	private String buildDeleteModelRequestBody(final String modelName) {
		final Map<String, Object> paramMap = new HashMap<>();
		paramMap.put("name", modelName);

		return JSONUtil.toJsonStr(paramMap);
	}

	/**
	 * 发送DELETE请求
	 *
	 * @param endpoint 请求端点
	 * @param paramJson 请求参数JSON
	 * @return 响应结果
	 */
	private HttpResponse sendDeleteRequest(String endpoint, String paramJson) {
		try {
			return HttpRequest.delete(config.getApiUrl() + endpoint)
				.header(Header.CONTENT_TYPE, "application/json")
				.header(Header.ACCEPT, "application/json")
				.body(paramJson)
				.timeout(config.getTimeout())
				.execute();
		} catch (Exception e) {
			throw new AIException("Failed to send DELETE request: " + e.getMessage(), e);
		}
	}

	// 构建copyModel请求体
	private String buildCopyModelRequestBody(final String source, final String destination) {
		Map<String, Object> requestBody = new HashMap<>();
		requestBody.put("source", source);
		requestBody.put("destination", destination);
		return JSONUtil.toJsonStr(requestBody);
	}

}

File: src/WellTool.AI/Model/Ollama/OllamaServiceImpl.cs
// Copyright (c) 2025 WellTool Team
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;
using WellTool.AI.Core;

namespace WellTool.AI.Model.Ollama
{
    /// <summary>
    /// Ollama服务实现类
    /// </summary>
    public class OllamaServiceImpl : BaseAIService, OllamaService
    {
        /// <summary>
        /// 构造
        /// </summary>
        /// <param name="config">AI配置</param>
        public OllamaServiceImpl(AIConfig config) : base(config) { }

        /// <summary>
        /// 对话
        /// </summary>
        /// <param name="prompt">user题词</param>
        /// <returns>AI回答</returns>
        public string Chat(string prompt)
        {
            var messages = new List<Message>
            {
                new Message("system", "You are a helpful assistant"),
                new Message("user", prompt)
            };
            return Chat(messages);
        }

        /// <summary>
        /// 对话-SSE流式输出
        /// </summary>
        /// <param name="prompt">user题词</param>
        /// <param name="callback">流式数据回调函数</param>
        public void Chat(string prompt, System.Action<string> callback)
        {
            var messages = new List<Message>
            {
                new Message("system", "You are a helpful assistant"),
                new Message("user", prompt)
            };
            Chat(messages, callback);
        }

        /// <summary>
        /// 对话
        /// </summary>
        /// <param name="messages">消息列表</param>
        /// <returns>AI回答</returns>
        public string Chat(List<Message> messages)
        {
            return ChatAsync(messages).GetAwaiter().GetResult();
        }

        /// <summary>
        /// 对话-SSE流式输出
        /// </summary>
        /// <param name="messages">消息列表</param>
        /// <param name="callback">回调函数</param>
        public void Chat(List<Message> messages, System.Action<string> callback)
        {
            ChatStreamAsync(messages, callback).GetAwaiter().GetResult();
        }

        /// <summary>
        /// 列出所有模型列表
        /// </summary>
        /// <returns>model列表</returns>
        public string Models()
        {
            return ModelsAsync().GetAwaiter().GetResult();
        }

        /// <summary>
        /// 拉取模型
        /// </summary>
        /// <param name="model">模型名称</param>
        /// <returns>拉取结果</returns>
        public string PullModel(string model)
        {
            return PullModelAsync(model).GetAwaiter().GetResult();
        }

        /// <summary>
        /// 删除模型
        /// </summary>
        /// <param name="model">模型名称</param>
        /// <returns>删除结果</returns>
        public string DeleteModel(string model)
        {
            return DeleteModelAsync(model).GetAwaiter().GetResult();
        }

        /// <summary>
        /// 异步对话
        /// </summary>
        /// <param name="messages">消息列表</param>
        /// <returns>AI回答</returns>
        private async Task<string> ChatAsync(List<Message> messages)
        {
            var requestBody = new
            {
                model = config.GetModel(),
                messages = messages.Select(m => new { role = m.Role, content = m.Content }).ToList(),
                temperature = 0.7
            };

            var json = JsonSerializer.Serialize(requestBody);
            var response = await SendPostAsync("/api/chat", json);
            return response;
        }

        /// <summary>
        /// 异步对话-SSE流式输出
        /// </summary>
        /// <param name="messages">消息列表</param>
        /// <param name="callback">回调函数</param>
        private async Task ChatStreamAsync(List<Message> messages, System.Action<string> callback)
        {
            var requestBody = new
            {
                model = config.GetModel(),
                messages = messages.Select(m => new { role = m.Role, content = m.Content }).ToList(),
                temperature = 0.7,
                stream = true
            };

            var json = JsonSerializer.Serialize(requestBody);
            await SendPostStreamAsync("/api/chat", json, callback);
        }

        /// <summary>
        /// 异步获取模型列表
        /// </summary>
        /// <returns>model列表</returns>
        private async Task<string> ModelsAsync()
        {
            return await SendGetAsync("/api/tags");
        }

        /// <summary>
        /// 异步拉取模型
        /// </summary>
        /// <param name="model">模型名称</param>
        /// <returns>拉取结果</returns>
        private async Task<string> PullModelAsync(string model)
        {
            var requestBody = new
            {
                name = model
            };

            var json = JsonSerializer.Serialize(requestBody);
            var response = await SendPostAsync("/api/pull", json);
            return response;
        }

        /// <summary>
        /// 异步删除模型
        /// </summary>
        /// <param name="model">模型名称</param>
        /// <returns>删除结果</returns>
        private async Task<string> DeleteModelAsync(string model)
        {
            var requestBody = new
            {
                name = model
            };

            var json = JsonSerializer.Serialize(requestBody);
            var response = await SendPostAsync("/api/delete", json);
            return response;
        }
    }
}
